app/src/main/java/org/mozilla/fenix/settings/address/controller/AddressEditorController.kt

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.settings.address.controller
+
+import androidx.navigation.NavController
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.concept.storage.UpdatableAddressFields
+import mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorage
+import org.mozilla.fenix.settings.address.AddressEditorFragment
+import org.mozilla.fenix.settings.address.interactor.AddressEditorInteractor
+
+/**
+ * [AddressEditorFragment] controller. An interface that handles the view manipulation of the
+ * credit card editor.
+ */
+interface AddressEditorController {
+
+    /**
+     * @see [AddressEditorInteractor.onCancelButtonClicked]
+     */
+    fun handleCancelButtonClicked()
+
+    /**
+     * @see [AddressEditorInteractor.onSaveAddress]
+     */
+    fun handleSaveAddress(addressFields: UpdatableAddressFields)
+}
+
+/**
+ * The default implementation of [AddressEditorController].
+ *
+ * @param storage An instance of the [AutofillCreditCardsAddressesStorage] for adding and retrieving
+ * addresses.
+ * @param lifecycleScope [CoroutineScope] scope to launch coroutines.
+ * @param navController [NavController] used for navigation.
+ * @param ioDispatcher [CoroutineDispatcher] used for executing async tasks. Defaults to [Dispatchers.IO].
+ */
+class DefaultAddressEditorController(
+    private val storage: AutofillCreditCardsAddressesStorage,
+    private val lifecycleScope: CoroutineScope,
+    private val navController: NavController,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+) : AddressEditorController {
+
+    override fun handleCancelButtonClicked() {
+        navController.popBackStack()
+    }
+
+    override fun handleSaveAddress(addressFields: UpdatableAddressFields) {
+        lifecycleScope.launch(ioDispatcher) {
+            storage.addAddress(addressFields)
+
+            lifecycleScope.launch(Dispatchers.Main) {
+                navController.popBackStack()
+            }
+        }
+    }
```

Comment:
It looks like the storage layer already handles switching to an IO dispatcher, so I don't think it's a necessary property for this class.

IMO it's best in general to handle context switching as close in the stack to the actual operation that requires it. This is evidenced in several popular libraries like Room and Retrofit, which will automatically handle moving to an appropriate dispatcher. It's talked about a bit in [this article about Room](https://medium.com/androiddevelopers/room-coroutines-422b786dc4c5). It allows us to put off the expensive operation of thread switching until the last possible moment, in case it's not actually needed.

Reply:
@MatthewTighe Can you illustrate what your suggested changes should look like? Are we just dropping the `ioDispatcher` here or is the proposal to change this function to a `suspend` function?

Reply:
I think you can just drop the `ioDispatcher` in this case.